rand_core/src/impls.rs

```diff
@@ -58,25 +58,25 @@ macro_rules! fill_via_chunks {
         let chunk_size_u8 = min($src.len() * SIZE, $dst.len());
         let chunk_size = (chunk_size_u8 + SIZE - 1) / SIZE;
 
-        // The following can be replaced with safe code, but unfortunately it's
-        // ca. 8% slower.
         if cfg!(target_endian = "little") {
+            // On LE we can do a simple copy, which is 25-50% faster:
             unsafe {
                 core::ptr::copy_nonoverlapping(
                     $src.as_ptr() as *const u8,
                     $dst.as_mut_ptr(),
                     chunk_size_u8);
             }
```

Comment:
You could also try the following safe approach. It adds some up-front bounds checks (which the compiler might be able to optimize out), but I would guess that it won't be a noticeable performance hit:

```rust
let src_u8_slice = zerocopy::AsBytes::as_bytes($src);
(&mut $dst[..chunk_size_u8]).copy_from_slice(&src_u8_slice[..chunk_size_u8]);
```

Reply:
Isn't this just trading one `unsafe` for another? We try to minimise `rand`'s dependency count (people have complained multiple times in the past). Another point of compromise :-(

Reply:
Technically, yes, although zerocopy's entire reason for existing is to encapsulate these `unsafe`s so that other crates don't have to think about it. The more crates that use zerocopy instead of using `unsafe`, the fewer instances of `unsafe` the ecosystem has to get right. Plus, when the [Lang Item for Transmutability](https://github.com/rust-lang/compiler-team/issues/411) is implemented and stabilized, `AsBytes` will be modified to just be a thin wrapper around those utilities, and will no longer require any `unsafe` itself.

Reply:
I'm still not keen on so many extra dependencies just for this one use of `unsafe`:
```
rand_core v0.6.3 (/home/dhardy/projects/rand/rand/rand_core)
├── getrandom v0.2.3
│   ├── cfg-if v1.0.0
│   └── libc v0.2.101
├── serde v1.0.130
│   └── serde_derive v1.0.130 (proc-macro)
│       ├── proc-macro2 v1.0.29
│       │   └── unicode-xid v0.2.2
│       ├── quote v1.0.9
│       │   └── proc-macro2 v1.0.29 (*)
│       └── syn v1.0.76
│           ├── proc-macro2 v1.0.29 (*)
│           ├── quote v1.0.9 (*)
│           └── unicode-xid v0.2.2
└── zerocopy v0.6.0
    ├── byteorder v1.4.3
    └── zerocopy-derive v0.3.0 (proc-macro)
        ├── proc-macro2 v1.0.29 (*)
        ├── syn v1.0.76 (*)
        └── synstructure v0.12.5
            ├── proc-macro2 v1.0.29 (*)
            ├── quote v1.0.9 (*)
            ├── syn v1.0.76 (*)
            └── unicode-xid v0.2.2
```
That's four extra deps when using `serde` (optional), nine when not, in what is essentially just an API crate. I know there's a valid argument that we shouldn't care, but (1) we've had complaints: #713, #850 and (2) if any of those happen to want to use `rand` (and I note that `zerocopy` does as a dev-dependency), then you have circular dependencies.

Reply:
`zerocopy-derive` looks like it would increase compile time significantly.

Reply:
> just for this one use of `unsafe`

I've identified a few other places that could use zerocopy:

- https://github.com/rust-random/rand/blob/master/src/rng.rs#L352-L357
- https://github.com/rust-random/rand/blob/master/src/rng.rs#L370-L375 (would require zerocopy to impl `FromBytes` and `AsBytes` for `core::num::Wrapping`)
- https://github.com/rust-random/rand/blob/master/src/distributions/integer.rs#L131-L135 (would require `packed_simd::Simd: FromBytes + AsBytes`, which isn't currently the case)

> That's four extra deps when using `serde` (optional)

Which four? I count one (`synstructure v0.12.5`) that is used transitively by `zerocopy-derive` and not transitively by `serde-derive`.

I mention these only to provide support for using zerocopy (and most importantly for getting rid of `unsafe` - if another crate like `bytemuck` is a better tradeoff, you should go with that instead), not to suggest that the tradeoff is definitely worth it.

Reply:
`bytemuck` looks like a better option at the moment, because the derive support is optional, so we could disable it and avoid the dependency / compile-time overhead. I think depending on `bytemuck` might be acceptable to reduce the unsafe code in Rand.